airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py

```diff
@@ -71,6 +75,66 @@ def get_checkpoint(self) -> Optional[Mapping[str, Any]]:
         return self._state
 
 
+class CursorBasedCheckpointReader(CheckpointReader):
+    """
+    CursorBasedCheckpointReader is used by streams that implement a Cursor in order to manage state. This allows the checkpoint
+    reader to delegate the complexity of fetching state to the cursor and focus on the iteration over a stream's partitions.
+    Right now only low-code connectors provide cursor implementations, but the logic is extensible to any stream that adheres
+    to the Cursor interface.
+    """
+
+    def __init__(self, cursor: Cursor, stream_slices: Iterable[Optional[Mapping[str, Any]]], read_state_from_cursor: bool = False):
+        # The first attempt of an RFR stream has an empty {} incoming state, but should still make a first attempt to read records
+        # from the first page in next().
+        self._cursor = cursor
+        self._stream_slices = iter(stream_slices)
+        self._read_state_from_cursor = read_state_from_cursor
+        self._current_slice: Optional[StreamSlice] = None
+        self._finished_sync = False
+
+    def next(self) -> Optional[Mapping[str, Any]]:
+        try:
+            if self._current_slice is None:
+                self._current_slice = self._get_next_slice()
+                return self._current_slice
+            if self._read_state_from_cursor:
+                state_for_slice = self._cursor.select_state(self._current_slice.get("partition"))
```

Comment:
nit: I think it's preferable to call `self._current_slice.partition` instead of get as slowly gets us away from thinking of the slice as an arbitrary dict.

Reply:
actually i realize it should just be `self._current_slice` since `select_state()` takes in the slice itself, not the partition. But regardless this is fixed and we don't reference the arbitrary dict. Thanks!